src/lang/primitive.cpp

```diff
@@ -248,8 +249,8 @@ namespace occa {
     case primitiveType::uint32_ : return primitive(!p.value.uint32_);
     case primitiveType::int64_  : return primitive(!p.value.int64_);
     case primitiveType::uint64_ : return primitive(!p.value.uint64_);
-    case primitiveType::float_  : return primitive(!p.value.float_);
-    case primitiveType::double_ : return primitive(!p.value.double_);
+    case primitiveType::float_  : OCCA_FORCE_ERROR("Cannot apply operator ! to float type");   break;
```

Comment:
These might be potentially used int he preprocessor, so we can't take them away

Reply:
Ok, I can do something like `return primitive(areBitwiseEqual(p.value.float_, 0))`?
Is it just this one, or do you want similar changes for `&&` and `||`? Since `&` and `|` are already disallowed in the current implementation I would say no?

Reply:
Maybe for these operations, we can first cast the `float`/`double` as `int`

Reply:
Ha ha, for float values, what is "zero" anyway?

My preference is the "...cannot apply operator ! to float type". Just leave it up to the user to explore the possibilities when a float type is involved based on his program logic. The user should already know he is in trouble if he's thinking about using logical not with a float value!

Yet I can't help but raise the question, for this primitive (`not`), is a sort of "near-`!`" acceptable, which checks for zero within _macheps_? e.g.,

`case primitiveType::float_  : return primitive( ((float)1.0 + p.value.float_) == (float)1.0 )`

My view: this is **not** acceptable for general use cases at this level. But others may disagree.

FYI possibly another approach to `!` based on an equality to zero comparison without employing the explicit `==` (i.e., _partial-order_ trick):

`case primitiveType::float_  : return primitive( !(p.value.float_ < -p.value.float_ || -p.value.float_ < p.value.float_) )`

All the above _mutatis mutandis_ for the double case.

Reply:
The argument made here is that we want to support as much C/C++ as possible, regardless of whether it's considered "good practice" or "bad practice". I 100% agree we shouldn't be doing exact comparisons with `float` or `double` "in practice". However, the goal here is language coverage.

---

TIL: Looks like floats aren't supported in the preprocessor!

```
main.cpp:4:6: error: floating constant in preprocessor expression
 #if !0.0
```

Reply:
Since the preprocessor won't accept them anyway, can we keep these operators unimplemented for float/double? That would still be my preferred solution.

Reply:
Sounds good!